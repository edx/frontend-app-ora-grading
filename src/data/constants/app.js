import { getConfig } from '@edx/frontend-platform';

export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
export const locationId = window.location.pathname.slice(1);
