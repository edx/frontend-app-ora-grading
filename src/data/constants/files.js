import { StrictDict } from 'utils';

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
});

export default FileTypes;
