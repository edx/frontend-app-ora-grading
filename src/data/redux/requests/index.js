export { actions, reducer } from './reducer';
