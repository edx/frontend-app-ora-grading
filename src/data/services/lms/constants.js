import { StrictDict } from 'utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const gradingStatusDisplay = StrictDict({
  [gradingStatuses.ungraded]: 'Ungraded',
  [gradingStatuses.locked]: 'Currently being graded by someone else',
  [gradingStatuses.graded]: 'Grading Complete',
  [gradingStatuses.inProgress]: 'You are currently grading this response',
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});
