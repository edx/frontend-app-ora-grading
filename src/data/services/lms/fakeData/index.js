import submissions from './submissionList';
import { mockSubmission } from './submissionFull';
import oraMetadata from './ora';
import ids from './ids';

export default {
  submissions,
  oraMetadata,
  mockSubmission,
  ids,
};
