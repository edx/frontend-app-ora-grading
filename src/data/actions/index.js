import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
