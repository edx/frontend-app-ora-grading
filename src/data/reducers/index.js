import { combineReducers } from 'redux';

import app from './app';
import submissions from './submissions';

/* istanbul ignore next */
const rootReducer = combineReducers({
  app,
  submissions,
});

export default rootReducer;
