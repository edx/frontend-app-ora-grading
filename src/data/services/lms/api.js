import { StrictDict } from 'utils';
import { gradeStatuses, lockStatuses } from './constants';
import fakeData from './fakeData';

// import urls from './urls';
// import { pageSize, paramKeys } from './constants';
// import messages from './messages';
// import * as utils from './utils';
// const { get, post, stringifyUrl } = utils;

/*********************************************************************************
 * GET Actions
 *********************************************************************************/

const mockSuccess = (returnValFn) => (...args) => (
  new Promise((resolve) => resolve(returnValFn(...args)))
);

const mockFailure = (returnValFn) => (...args) => (
  new Promise((resolve, reject) => reject(returnValFn(...args)))
);

/**
 * get('/api/initialize', { ora_location, course_id? })
 * @return {
 *   oraMetadata: { name, prompt, type ('individual' vs 'team'), rubricConfig, fileUploadResponseConfig },
 *   courseMetadata: { courseOrg, courseName, courseNumber, courseId },
 *   submissions: {
 *     [submissionUUID]: {
 *       id: <submissionUUID>, (not currently used)
 *       username
 *       submissionUUID
 *       dateSubmitted (timestamp)
 *       gradeStatus (['ungraded', 'graded', 'locked', 'locked_by_you'?])
 *       grade: { pointsEarned, pointsPossible }
 *     },
 *     ...
 *   },
 * }
 */
const initializeApp = mockSuccess(() => ({
  oraMetadata: fakeData.oraMetadata,
  courseMetadata: fakeData.courseMetadata,
  submissions: fakeData.submissions,
}));

/**
 * get('/api/submission', { submissionUUID })
 * @return {
 *   submision: {
 *     gradeData,
 *     gradeStatus,
 *     response: { files: [{}], text: <html> },
 *   },
 * }
 */
const fetchSubmission = mockSuccess((submissionUUID) => {
  console.log({ fetchSubmission: { submissionUUID } });
  return fakeData.mockSubmission(submissionUUID);
});

/**
 * fetches the current grade, gradeStatus, and rubricResponse data for the given submission
 * get('/api/submissionStatus', { submissionUUID })
 *   @return {obj} submissionStatus object
 *   {
 *     gradeData,
 *     gradeStatus,
 *     lockStatus,
 *   }
 */
const fetchSubmissionStatus = mockSuccess((submissionUUID) => (
  fakeData.mockSubmissionStatus(submissionUUID)
));

/**
 * Fetches only the learner response for a given submission. Used for pre-fetching response
 * for neighboring submissions in the queue.
 */
export const fetchSubmissionResponse = mockSuccess((submissionUUID) => ({
  response: fakeData.mockSubmission(submissionUUID).response,
}));

/* I assume this is the "Start Grading" call, even for if a
 * submission is already graded and we are attempting re-lock.
 * Assuming the check for if allowed would happen locally first.
 * post('api/lock', { ora_location, submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmission = mockSuccess(({ submissionUUID, value }) => ({
  ...fakeData.mockSubmissionStatus(submissionUUID),
  lockStatus: value ? lockStatuses.inProgress : lockStatuses.unlocked,
}));

/*
 * Assuming we do not care who has locked it or why, as there
 * is no design around communicating that info
 * post('api/lock', { submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmissionFail = mockFailure(() => ({
  error: 'that did not work',
}));

/*
 * post('api/updateGrade', { submissionUUID, gradeData })
 * @param {object} gradeData - full grading submission data
 */
const updateGrade = mockSuccess((submissionUUID, gradeData) => ({
  gradeData,
  gradeStatus: gradeStatuses.graded,
  lockStatus: lockStatuses.unlocked,
}));

export default StrictDict({
  initializeApp,
  fetchSubmission,
  fetchSubmissionResponse,
  fetchSubmissionStatus,
  lockSubmission,
  lockSubmissionFail,
  updateGrade,
});
