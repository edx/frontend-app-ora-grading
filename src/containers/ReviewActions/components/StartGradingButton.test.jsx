import React from 'react';
import { shallow } from 'enzyme';

import { selectors, thunkActions } from 'data/redux';
import { gradingStatuses as statuses } from 'data/services/lms/constants';

import {
  StartGradingButton,
  mapStateToProps,
  mapDispatchToProps,
} from './StartGradingButton';

jest.mock('data/redux/grading/selectors', () => ({
  selected: {
    gradeStatus: (state) => ({ gradeStatus: state }),
    gradingStatus: (state) => ({ gradingStatus: state }),
  },
}));
jest.mock('./OverrideGradeConfirmModal', () => 'OverrideGradeConfirmModal');
jest.mock('./StopGradingConfirmModal', () => 'StopGradingConfirmModal');

let el;

describe('StartGradingButton component', () => {
  describe('component', () => {
    const props = {};
    beforeEach(() => {
      props.startGrading = jest.fn().mockName('this.props.startGrading');
      props.stopGrading = jest.fn().mockName('this.props.stopGrading');
    });
    describe('snapshotes', () => {
      const mockedEl = (gradingStatus, gradeStatus) => {
        const renderedEl = shallow(
          <StartGradingButton
            {...props}
            gradingStatus={gradingStatus}
            gradeStatus={gradeStatus || gradingStatus}
          />,
        );
        const mockMethod = (methodName) => {
          renderedEl.instance()[methodName] = jest.fn().mockName(`this.${methodName}`);
        };
        mockMethod('handleClick');
        mockMethod('hideConfirmOverrideGrade');
        mockMethod('confirmOverrideGrade');
        mockMethod('hideConfirmStopGrading');
        mockMethod('confirmStopGrading');
        return renderedEl;
      };
      test('snapshot: locked (null)', () => {
        el = mockedEl(statuses.locked);
        expect(el.instance().render()).toMatchSnapshot();
        expect(el.isEmptyRender()).toEqual(true);
      });
      test('snapshot: ungraded (startGrading callback)', () => {
        expect(mockedEl(statuses.ungraded).instance().render()).toMatchSnapshot();
      });
      test('snapshot: graded, confirmOverride (startGrading callback)', () => {
        el = mockedEl(statuses.graded);
        el.setState({ showConfirmOverrideGrade: true });
        expect(el.instance().render()).toMatchSnapshot();
      });
      test('snapshot: inProgress, isOverride, confirmStop (stopGrading callback)', () => {
        el = mockedEl(statuses.inProgress, statuses.graded);
        el.setState({ showConfirmStopGrading: true });
        expect(el.instance().render()).toMatchSnapshot();
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { some: 'test-state' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('gradeStatus loads from grading.selected.gradeStatus', () => {
      expect(mapped.gradeStatus).toEqual(selectors.grading.selected.gradeStatus(testState));
    });
    test('gradingStatus loads from grading.selected.gradingStatus', () => {
      expect(mapped.gradingStatus).toEqual(selectors.grading.selected.gradingStatus(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    it('loads startGrading from thunkActions.grading.stargGrading', () => {
      expect(mapDispatchToProps.startGrading).toEqual(thunkActions.grading.startGrading);
    });
    it('loads stopGrading from thunkActions.grading.stopGrading', () => {
      expect(mapDispatchToProps.stopGrading).toEqual(thunkActions.grading.stopGrading);
    });
  });
});
