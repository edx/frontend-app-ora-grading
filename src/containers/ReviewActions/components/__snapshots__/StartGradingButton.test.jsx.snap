// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartGradingButton component component snapshotes snapshot: grade pending (disabled) 1`] = `
<React.Fragment>
  <Button
    disabled={true}
    iconAfter={[MockFunction icons.Highlight]}
    onClick={[MockFunction this.handleClick]}
    variant="primary"
  >
    <FormattedMessage
      defaultMessage="Start grading"
      description="Review pane button text to start grading"
      id="ora-grading.ReviewActions.StartGradingButton.startGrading"
    />
  </Button>
  <OverrideGradeConfirmModal
    isOpen={false}
    onCancel={[MockFunction this.hideConfirmOverrideGrade]}
    onConfirm={[MockFunction this.confirmOverrideGrade]}
  />
  <StopGradingConfirmModal
    isOpen={false}
    isOverride={false}
    onCancel={[MockFunction this.hideConfirmStopGrading]}
    onConfirm={[MockFunction this.confirmStopGrading]}
  />
</React.Fragment>
`;

exports[`StartGradingButton component component snapshotes snapshot: graded, confirmOverride (startGrading callback) 1`] = `
<React.Fragment>
  <Button
    disabled={false}
    iconAfter={[MockFunction icons.Highlight]}
    onClick={[MockFunction this.handleClick]}
    variant="primary"
  >
    <FormattedMessage
      defaultMessage="Override grade"
      description="Review pane button text to start grading an already graded submission"
      id="ora-grading.ReviewActions.StartGradingButton.overrideGrade"
    />
  </Button>
  <OverrideGradeConfirmModal
    isOpen={true}
    onCancel={[MockFunction this.hideConfirmOverrideGrade]}
    onConfirm={[MockFunction this.confirmOverrideGrade]}
  />
  <StopGradingConfirmModal
    isOpen={false}
    isOverride={true}
    onCancel={[MockFunction this.hideConfirmStopGrading]}
    onConfirm={[MockFunction this.confirmStopGrading]}
  />
</React.Fragment>
`;

exports[`StartGradingButton component component snapshotes snapshot: inProgress, isOverride, confirmStop (stopGrading callback) 1`] = `
<React.Fragment>
  <Button
    disabled={false}
    iconAfter={[MockFunction icons.Cancel]}
    onClick={[MockFunction this.handleClick]}
    variant="primary"
  >
    <FormattedMessage
      defaultMessage="Stop grading this response"
      description="Review pane button text to stop grading"
      id="ora-grading.ReviewActions.StartGradingButton.stopGrading"
    />
  </Button>
  <OverrideGradeConfirmModal
    isOpen={false}
    onCancel={[MockFunction this.hideConfirmOverrideGrade]}
    onConfirm={[MockFunction this.confirmOverrideGrade]}
  />
  <StopGradingConfirmModal
    isOpen={true}
    isOverride={true}
    onCancel={[MockFunction this.hideConfirmStopGrading]}
    onConfirm={[MockFunction this.confirmStopGrading]}
  />
</React.Fragment>
`;

exports[`StartGradingButton component component snapshotes snapshot: lock pending (disabled) 1`] = `
<React.Fragment>
  <Button
    disabled={true}
    iconAfter={[MockFunction icons.Highlight]}
    onClick={[MockFunction this.handleClick]}
    variant="primary"
  >
    <FormattedMessage
      defaultMessage="Start grading"
      description="Review pane button text to start grading"
      id="ora-grading.ReviewActions.StartGradingButton.startGrading"
    />
  </Button>
  <OverrideGradeConfirmModal
    isOpen={false}
    onCancel={[MockFunction this.hideConfirmOverrideGrade]}
    onConfirm={[MockFunction this.confirmOverrideGrade]}
  />
  <StopGradingConfirmModal
    isOpen={false}
    isOverride={false}
    onCancel={[MockFunction this.hideConfirmStopGrading]}
    onConfirm={[MockFunction this.confirmStopGrading]}
  />
</React.Fragment>
`;

exports[`StartGradingButton component component snapshotes snapshot: locked (null) 1`] = `null`;

exports[`StartGradingButton component component snapshotes snapshot: ungraded (startGrading callback) 1`] = `
<React.Fragment>
  <Button
    disabled={false}
    iconAfter={[MockFunction icons.Highlight]}
    onClick={[MockFunction this.handleClick]}
    variant="primary"
  >
    <FormattedMessage
      defaultMessage="Start grading"
      description="Review pane button text to start grading"
      id="ora-grading.ReviewActions.StartGradingButton.startGrading"
    />
  </Button>
  <OverrideGradeConfirmModal
    isOpen={false}
    onCancel={[MockFunction this.hideConfirmOverrideGrade]}
    onConfirm={[MockFunction this.confirmOverrideGrade]}
  />
  <StopGradingConfirmModal
    isOpen={false}
    isOverride={false}
    onCancel={[MockFunction this.hideConfirmStopGrading]}
    onConfirm={[MockFunction this.confirmStopGrading]}
  />
</React.Fragment>
`;
